pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java

```diff
@@ -89,12 +92,28 @@ protected IntermediateResultsBlock getNextBlock() {
       _futures[i] = _executorService.submit(new TraceRunnable() {
         @Override
         public void runJob() {
+          ThreadTimer processThreadTimer = new ThreadTimer();
+          processThreadTimer.start();
+
           processSegments(threadIndex);
+
+          processThreadTimer.stop();
+          totalWorkerTime.addAndGet(processThreadTimer.getThreadTime());
         }
       });
     }
 
+    ThreadTimer mergeThreadTimer = new ThreadTimer();
```

Comment:
I don't think we should instrument the merge part this way.

We should do that in InstanceResponseOperator and InstanceResponseBlock because that is the single main thread and all the merge work is happening under that thread. 

So we start the main thread timer in InstanceResponseOperator and stop the main thread timer in InstanceResponseBlock's constructor and then add up the timer received from merged intermediate result block. This will cover both single threaded and multi-threaded code.

Reply:
If we instrument in that way (start the main thread timer in InstanceResponseOperator), then the total time will account an extra processing thread time interval:

example:
t0: InstanceResponseOperator call InstanceResponseBlock's constructor, which will call _operator.nextBlock();
t2: process segments in parallel. let's say we have 5 thread processing segments in parallel. Let's assume all the 5 thread start and end at the same time.
t3: processing end, start merge
t4: merge end.

NOTE: since time spending on constructor is negligible, let's assume t0=t2.

Then, expected_totalTime = processing_time + merge_time =  (t3-t2)*5 + (t4-t3). However, if start the main thread timer in InstanceResponseOperator, actual_totalTime = t4-t0+(t3-t2)*5 = t4-t2+(t3-t2)*5

actual_totalTime - expected_totalTime = (t4-t2)-(t4-t3) = t3-t2, in other word, there is an extra processing thread time get accounted.

I will let @mcvsubbu chime in to make a double check.

Reply:
As discussed offline, we should move all the single threaded instrumentation to InstanceResponseOperator

```
 @Override
  protected InstanceResponseBlock getNextBlock() {
    ThreadTimer mainThreadTimer = new ThreadTimer();
   
    mainThreadTimer.start();

    IntermediateResultsBlock intermediateResultsBlock = (IntermediateResultsBlock) _operator.nextBlock();
    InstanceResponseBlock instanceResponseBlock = new InstanceResponseBlock(intermediateResultsBlock);
    DataTable dataTable = instanceResponseBlock.getInstanceResponseDataTable();
   
    mainThreadTimer.stop();
   
    long totalThreadCpuTimeNs = intermediateResultsBlock.getThreadCpuTimeNs() + mainThreadTimer.getThreadTimeNs();
   
    dataTable.getMetadata().put(DataTable.THREAD_CPU_TIME_NS_METADATA_KEY, String.valueOf(totalThreadCpuTimeNs));

    return instanceResponseBlock
  }
```

This way all single threaded query execution (except for planning) code's cpu time is covered without having to do it in multiple places.

Reply:
cc @mcvsubbu @mqliang